File: Leetcode/1992/1992.cpp
constexpr int STEPS = 2;
const int dx[] = {0,1};
const int dy[] = {1,0};

namespace DSU {
    vector<int> father;
    vector<int> Size;
    void init(int n) {
        vector<int>().swap(father);
        vector<int>().swap(Size);
        father.resize(n + 5);
        iota(father.begin(), father.end(), 0);
        Size.resize(n + 5, 1);
    }
    int find(int x) {
        if (x != father[x]) {
            father[x] = find(father[x]);
        }
        return father[x];
    }
    void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return;
        }
        if (Size[x] >
            Size[y]) {
            swap(x, y);
        }
        father[x] = y;
        Size[y] += Size[x];
    }
}

class Solution {
public:
    vector<vector<int>> findFarmland(vector<vector<int>>& land) {
        int n = (int)land.size();
        int m = (int)land[0].size();
        DSU::init(n*m);
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (land[i][j] == 1) {
                    int new_i,new_j;
                    for (int step = 0; step < STEPS; ++step) {
                        new_i = i+dx[step];
                        new_j = j+dy[step];
                        if (new_i >= 0 && new_i < n && new_j >= 0 && new_j < m && land[new_i][new_j] == 1) {
                            DSU::merge(new_i*m+new_j,i*m+j);
                        }
                    }
                }
            }
        }
        unordered_map<int,vector<pair<int,int>>> unmap;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (land[i][j] == 1) {
                    unmap[DSU::find(i*m+j)].emplace_back(i,j);
                }
            }
        }
        vector<vector<int>> ans;
        for (auto &[k,v] : unmap) {
            auto iter = minmax_element(v.begin(),v.end());
            ans.emplace_back(vector<int>{iter.first->first,iter.first->second,iter.second->first,iter.second->second});
        }
        return ans;
    }
};

File: Leetcode/1992/1992.go
const STEPS = 2
var dx = [STEPS]int{0,1}
var dy = [STEPS]int{1,0}

var (
	father []int
	Size []int
)
func dsuInit(n int) {
	father = make([]int, n+5)
	Size = make([]int, n+5)
	for i := 0; i < len(father); i++ {
		father[i] = i
	}
	for i := 0; i < len(Size); i++ {
		Size[i] = 1
	}
}
func find(x int) int {
	if father[x] != x {
		father[x] = find(father[x])
	}
	return father[x]
}
func merge(x, y int) {
	x = find(x)
	y = find(y)
	if x == y {
		return
	}
	if Size[x] > Size[y] {
		x, y = y, x
	}
	Size[y] += Size[x]
	father[x] = y
}

func minmax(s []int) (int,int) {
	minVal, maxVal := s[0], s[0]
	for i := 1; i < len(s); i++ {
		if minVal > s[i] {
			minVal = s[i]
		}
		if maxVal < s[i] {
			maxVal = s[i]
		}
	}
	return minVal, maxVal
}

func findFarmland(land [][]int) [][]int {
	n := len(land)
	m := len(land[0])
	dsuInit(n*m)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if land[i][j] == 1 {
				for step := 0; step < STEPS; step++ {
					newI := i + dx[step]
					newJ := j + dy[step]
					if newI >= 0 && newI < n && newJ >= 0 && newJ < m && land[newI][newJ] == 1 {
						merge(i*m+j,newI*m+newJ)
					}
				}
			}
		}
	}
	unmap := make(map[int][]int)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if land[i][j] == 1 {
				unmap[find(i*m+j)] = append(unmap[find(i*m+j)], i*m+j)
			}
		}
	}
	var ans [][]int
	for _, v := range unmap {
		minVal, maxVal := minmax(v)
		ans = append(ans, []int{minVal/m,minVal%m,maxVal/m,maxVal%m})
	}
	return ans
}